index.js

```diff
@@ -85,109 +81,138 @@ PersonaClient.prototype.validateToken = function (token, scope, overridingScope,
-                    next(null, VALIDATED_TYPES.PERSONA);
+    var headScopeThenVerify = function headScope(scope, callback) {
+        var options = {
+            hostname: this.config.persona_host,
+            port: this.config.persona_port,
+            path: this.config.persona_oauth_route + token + "?scope=" + scope,
+            method: "HEAD"
+        };
+
+        this.http.request(options, function onSuccess(response) {
+            switch(response.statusCode) {
+            case 204:
+                return callback(null, "ok");
+            case 401:
+                return callback(ERROR_TYPES.VALIDATION_FAILURE, null);
+            case 403:
+                if(scope === "su") {
+                    // We tried using su and can go no further
+                    return callback(ERROR_TYPES.INSUFFICIENT_SCOPE, null);
                 } else {
-                    if (usingScope && usingScope !== "su") {
-                        // try su, they are allowed to perform any operation
-                        _this.validateToken (token, "su", overridingScope, next);
-                    } else {
-                        _this.debug("Verification failed for token " + cacheKey + " with status code " + oauthResp.statusCode);
-                        if (oauthResp.statusCode === 403) {
-                            next(ERROR_TYPES.INSUFFICIENT_SCOPE, null);
-                        } else {
-                            next(ERROR_TYPES.VALIDATION_FAILURE, null);
-                        }
-                    }
+                    // Try again with su
+                    return headScopeThenVerify("su", callback);
                 }
-            }).on ("error", function (e) {
-                _this.error("OAuth::validateToken problem: " + e.message);
-                next(e.message, null);
-            }).end();
-        }
-    });
+                break;
+            default:
+                return callback(ERROR_TYPES.COMMUNICATION_ISSUE, null);
+            }
+        }).on("error", function onError(error) {
+            return callback(error, null);
+        }).end();
+    };
+    headScopeThenVerify = headScopeThenVerify.bind(this);
+
+    var verifyToken = function verifyToken(publicKey) {
+        // Make sure we cache the key if this function was called as the result of a HTTP request to fetch it
+        this.publicKey = publicKey;
+        var jwtConfig = {
+            algorithms: ["RS256"]
+        };
+
+        jwt.verify(token, publicKey, jwtConfig, function onVerify(error, decodedToken) {
+            if(error) {
+                return next(ERROR_TYPES.VALIDATION_FAILURE, null);
+            }
+
+            if(scope && decodedToken.hasOwnProperty("scopeCount")) {
+                // Too many scopes to put in the payload so we'll have to ask Persona
+                return headScopeThenVerify(scope, next);
+            } else if (!scope || _.includes(decodedToken.scopes, "su") || _.includes(decodedToken.scopes, scope)) {
```

Comment:
could `scope` be something silly like `"0"` or `0`? Maybe we should check for `scope === null || scope === undefined` instead.